File: Programmers/정권호/Q42583.java
package programers.lv2._0613;

import java.util.LinkedList;
import java.util.Queue;

public class Q42583 {
    class Solution {
        public int solution(int bridge_length, int weight, int[] truck_weights) {
            int answer = 1;
            int now = 0;
            int idx = 0;
            Queue<Integer[]> queue = new LinkedList<>();
            queue.add(new Integer[]{truck_weights[idx++], answer + bridge_length});
            now = truck_weights[0];

            while (!queue.isEmpty()) {
                answer++;
                Integer[] peek = queue.peek();
                if (answer == peek[1]) {
                    queue.poll();
                    now -= peek[0];
                }

                if (idx < truck_weights.length && weight >= now + truck_weights[idx]) {
                    if(idx == truck_weights.length - 1) {
                        answer += bridge_length;
                        break;
                    }
                    queue.add(new Integer[]{truck_weights[idx], answer + bridge_length});
                    now += truck_weights[idx];
                    idx++;

                }
            }
            return answer;
        }
    }
}

File: Programmers/유양우/Q42583.kt
import java.util.*


fun main() {
    solution(2, 10, intArrayOf(7, 4, 5, 6))
}

fun solution(bridge_length: Int, weight: Int, truck_weights: IntArray): Int {
    // bridge_lenth 다리에 올라갈 수 있는 트럭 수
    // weight 다리가 견딜 수 있는 무게
    // truck_weights 트럭별 무게
    var time = 0
    var waitQueue: Queue<Int> = LinkedList<Int>() // 대기중인 트럭
    var bridgeQueue: Queue<Int> = LinkedList<Int>() // 다리위에 있는 트럭
    truck_weights.forEach {
        waitQueue.add(it)
    }
    for (i in 0 until bridge_length) {
        bridgeQueue.add(0)
    }
    while (bridgeQueue.isNotEmpty()) {
        time++
        bridgeQueue.poll()
        if (waitQueue.isNotEmpty()) {
            if (bridgeQueue.sum() + waitQueue.peek() <= weight) {
                bridgeQueue.add(waitQueue.poll())
            } else {
                bridgeQueue.add(0)
            }
        }
    }

    print(time)

    return time
}
